pkg/scale/types.go

```diff
@@ -14,27 +14,42 @@
 
 package scale
 
-import "github.com/banzaicloud/koperator/api/v1beta1"
+import (
+	"github.com/banzaicloud/go-cruise-control/pkg/api"
+	"github.com/banzaicloud/go-cruise-control/pkg/types"
+
+	"github.com/banzaicloud/koperator/api/v1beta1"
+)
 
 type CruiseControlScaler interface {
 	IsReady() bool
-	Status() CruiseControlStatus
+	Status() (CruiseControlStatus, error)
 	GetUserTasks(taskIDs ...string) ([]*Result, error)
 	IsUp() bool
 	AddBrokers(brokerIDs ...string) (*Result, error)
+	AddBrokersWithParams(params map[string]string) (*Result, error)
+	RemoveBrokersWithParams(params map[string]string) (*Result, error)
+	RebalanceWithParams(params map[string]string) (*Result, error)
+	StopExecution() (*Result, error)
+	GetNumMonitoredWin() (float32, types.MonitorState, error)
 	RemoveBrokers(brokerIDs ...string) (*Result, error)
 	RebalanceDisks(brokerIDs ...string) (*Result, error)
 	BrokersWithState(states ...KafkaBrokerState) ([]string, error)
+	GetKafkaClusterState() (*types.KafkaClusterState, error)
 	PartitionReplicasByBroker() (map[string]int32, error)
 	BrokerWithLeastPartitionReplicas() (string, error)
 	LogDirsByBroker() (map[string]map[LogDirState][]string, error)
+	GetKafkaClusterLoad() (*api.KafkaClusterLoadResponse, error)
 }
 
 type Result struct {
-	TaskID    string
-	StartedAt string
-	State     v1beta1.CruiseControlUserTaskState
-	Err       string
+	TaskID             string
+	StartedAt          string
+	ResponseStatusCode int
+	RequestURL         string
+	Result             *types.OptimizationResult
+	State              v1beta1.CruiseControlUserTaskState
+	Err                string
```

Comment:
This error message is going to the CruiseControlOperation.CurrentTask.ErrorMessage (string)
I think it would better to use the error interface here and later stringify.

Reply:
I agree.
I would always advise postponing destructurization (converting highly structured data into lower level of structured data or unstructured data) as far as we can put it off to avoid loosing information which might come in handy later.

Reply:
You changed the references setting this but not the field type as I saw.